### PR TITLE
[hist] Add backwards-compatibility for THnSparseL typedef

### LINKED ISSUE #15874: [Hist] Backwards compatibility broken for `THnSparseL` in 6.32
### Check duplicate issues.

- [ ] Checked for duplicates

### Description

Backwards compatibility for `THnSparseL` was broken in 6.32 by #8546, @ferdymercury @vepadulano.

### Reproducer

Write a `THnSparseL` to a file e.g. with ROOT 6.30.06:
```c++
void write_thnsparse_l()
{
   std::unique_ptr<TFile> file{TFile::Open("hist.root", "RECREATE")};
   Int_t bins[2] = {10, 20};
   Double_t xmin[2] = {0., -5.};
   Double_t xmax[2] = {10., 5.};
   THnSparseL hist{"hist", "hist", 2, bins, xmin, xmax};
   file->WriteObject(&hist, "hist");
}
```
Try to read it with ROOT 6.32.00 (or `master`):
```bash
root hist.root -e 'hist'
```
You'll get this error:
```txt
Error in <TClingCallFunc::make_ctor_wrapper>: Failed to compile
  ==== SOURCE BEGIN ====
__attribute__((used)) extern "C" void __ctor_0(void** ret, void* arena, unsigned long nary)
{
   if (!arena) {
      if (!nary) {
         *ret = new THnSparseT<TArrayL>;
      }
      else {
         *ret = new THnSparseT<TArrayL>[nary];
      }
   }
   else {
      if (!nary) {
         *ret = new (arena) THnSparseT<TArrayL>;
      }
      else {
         *ret = new (arena) THnSparseT<TArrayL>[nary];
      }
   }
}

  ==== SOURCE END ====
Error in <TClingCallFunc::ExecDefaultConstructor>: Called with no wrapper, not implemented!
Error in <TClingClassInfo::New()>: Call of default constructor failed to return an object for class: THnSparseT<TArrayL>
Error in <TClass::New>: cannot create object of class THnSparseT<TArrayL>
Error in <TKey::ReadObj>: Cannot create new object of class THnSparseT<TArrayL>
IncrementalExecutor::executeFunction: symbol '_ZN10THnSparseTI7TArrayLE5ClassEv' unresolved while linking [cling interface function]!
You are probably missing the definition of THnSparseT<TArrayL>::Class()
Maybe you need to load the corresponding shared library?
IncrementalExecutor::executeFunction: symbol '_ZN10THnSparseTI7TArrayLE8StreamerER7TBuffer' unresolved while linking [cling interface function]!
You are probably missing the definition of THnSparseT<TArrayL>::Streamer(TBuffer&)
Maybe you need to load the corresponding shared library?
```

### ROOT version

ROOT 6.32.00 upwards.

### Additional context

Thanks do @philippe554 for catching this! This issue is important for project HighLO because we're using `THnSparseL` as a container for reduced data.

### PR DESCRIPTION
**Pull request by vepadulano**
Fixes #15874

### CODE REVIEW

**Review by dpiparo — APPROVED** [review]

**Comment by vepadulano** [comment]
Added backwards-compatibility test in roottest, will wait for the CI to run.

File: hist/hist/inc/LinkDef.h [hunk 1]
```diff
@@ -147,7 +147,9 @@
 #pragma link C++ class THnSparse+;
 #pragma link C++ class THnSparseT<TArrayD>+;
 #pragma link C++ class THnSparseT<TArrayF>+;
+#pragma link C++ class THnSparseT<TArrayL>+;
```
Comment (pcanal) [inline hist/hist/inc/LinkDef.h:150]: We might still be missing a deprecation warning.

Comment (vepadulano) [reply hist/hist/inc/LinkDef.h:150]: Where/how should I add that? We have so many ways to deprecate old classes, and this is not really deprecated to be fair.

**Comment by dpiparo** [comment]
Maybe we can proceed with the merge and backport to 6.32? This would prepare us to the next patch release...

**Comment by vepadulano** [comment]
Failures are unrelated, merging this PR due to its relevance for the upcoming patch release.

**Pull request merged**